### PR TITLE
UI: Source context menu

### PR DESCRIPTION
**Pull request by DDRBoxman**
Test Build:
https://dev.azure.com/obsjim/obsjim/_build/results?buildId=2733&view=artifacts&type=publishedArtifacts

Screen Shot 2020-02-24 at 12 01 22 AM
Screen Shot 2020-02-24 at 12 01 30 AM
Screen Shot 2020-02-24 at 12 07 14 AM

See #1447 for more history

TODO:
- [x] Implement a override for the icon render system to make the disabled ones darker (https://doc.qt.io/qt-5/qproxystyle.html)

### CODE REVIEW

**Comment by DDRBoxman** [comment]
https://stackoverflow.com/questions/30865189/change-disabled-qicon-tint-color

**Comment by WizardCM** [comment]
Friendly reminder: this PR has a lot of merge conflicts

**Comment by WizardCM** [comment]
So, I love this idea. Functionally, it works well.

Quirks I've run into with the latest code in the branch:
* Dock popout/close buttons are made much larger with these changes
* The Filters icon seems to be invisible & the Center icon seems off-center? Might just be my build environment
* Positioning & alignment buttons currently don't grey out for audio-only sources
* Renaming the actively selected source doesn't update its name in the context menu
* I think calling it a "Context Bar" may work better, purely because "context menu" is often used to describe a right click menu
* Most people are used to bars like this being along the top of an application, rather than somewhere near the bottom. I'm personally not sure whether it should be locked to the preview, or to just appear at the top of the window itself. **I don't think it should be a dock**, but maybe users should be able to right click it and choose some preset positions? "Top of window", "above preview", "below preview"?
* The separators between sections look like they'd be gripable/resizable (say, to fit a longer source name), but aren't. Users may notice this too
* I think it may be worth adding "Edit Transform" as one of the buttons as it'll make that capability significantly more discoverable



Other than these notes, I think it looks good, and provides a much easier way to reach Filters & common transform actions.

**Comment by cg2121** [comment]
Just tested it out. I love this feature.

Some feedback:
- When there are no sources selected, the menu should still be shown, just disabled.
- I would prefer it to be at the top, below the main menu.

**Comment by cg2121** [comment]
When the menu is disabled the source label probably should be "No source selected"

**Comment by DDRBoxman** [comment]
Added in text for the no source selected attn @cg2121 

Also went ahead and did a squash

**Comment by DDRBoxman** [comment]
https://dev.azure.com/obsjim/obsjim/_build/results?buildId=2723&view=artifacts&type=publishedArtifacts

File: UI/CMakeLists.txt [hunk 1]
```diff
@@ -325,7 +325,7 @@ set(obs_HEADERS
 	remote-text.hpp
 	audio-encoders.hpp
 	qt-wrappers.hpp
-	clickable-label.hpp)
+	clickable-label.hpp obs-proxy-style.cpp obs-proxy-style.hpp)
```
Comment (cg2121) [inline UI/CMakeLists.txt]: The cpp file shouldn't be in the headers section. Also the files should be on separate lines.

File: UI/CMakeLists.txt [hunk 2]
```diff
@@ -335,7 +335,7 @@ set(obs_importers_SOURCES
 	importers/classic.cpp
 	importers/sl.cpp
 	importers/studio.cpp
-	importers/xsplit.cpp)
+	importers/xsplit.cpp locked-checkbox.cpp visibility-checkbox.cpp)
```
Comment (cg2121) [inline UI/CMakeLists.txt]: Separate lines here as well.

**Comment by dodgepong** [comment]
Testing this out now, some feedback:

* I expected the Visibility and Lock icons to be buttons. Should they be buttons instead of just status indicators?
* I was able to shrink the preview this small: https://i.imgur.com/EBoIkwv.png When the preview gets that small, the context bar gets to be pretty unusable. Is there a way to make the context bar "responsive", such that buttons get collapsed into menus when the space gets too small to render the whole bar? This might also allow for users to have the bar along the left or right side of the preview rather than be limited to horizontal renders.
* That said, I don't know how a vertical render or a "responsive" layout would work if we were to still allow things like media controls. Do you have any ideas on how to handle cases where the preview is smaller than a given size?
* I think the icon for "Edit Transform" is a bit ambiguous, but I'm not sure off the top of my head what it should be instead. Maybe it should have a text label like Properties and Filters do?

**Comment by DDRBoxman** [comment]
hmm, the visibility and lost indicators should work the same as in the source list. Was the issue that they didn't highlight? They're checkboxes under the hood.

**Comment by dodgepong** [comment]
Clicking on them doesn't do anything for me at all.

**Comment by DDRBoxman** [comment]
@dodgepong just pushed some code to fix the lock / visible stuff not working. side effects from the new disable stuff.

I'll have to do some research but I have a feeling doing responsive stuff in QT is going to suck :(

**Comment by DDRBoxman** [comment]
https://dev.azure.com/obsjim/obsjim/_build/results?buildId=2733&view=artifacts&type=publishedArtifacts

**Comment by dodgepong** [comment]
Cool, that fixed the lock/visibility buttons.

Maybe for the responsive stuff, a first-pass could be that if it would be clipped, just hide the whole bar? Or collapse it to a menu that opens from one button? Again, not sure how something like media controls would work inside a menu -- maybe a collapsible floating panel that _acts_ like a menu? Not sure if that's possible in Qt, just thinking in HTML terms right now.

**Comment by COOLIGUAY** [comment]
I am testing with sources in groups.

When you select sources inside a group it does not work properly.

The first source you click is selected and does not change until you select the group again or another source outside it.

If you come from not having selected any source, remains until you select the group or other source outside it too.

3e8c35994a035f4919ee6370a9b655e5

**Comment by WizardCM** [comment]
I can confirm the bug @COOLIGUAY ran into. Selecting different sources within groups definitely doesn't update the context bar.

**Comment by DDRBoxman** [comment]
https://github.com/obsproject/obs-studio/issues/2794

**Comment by xiCO2k** [comment]
image
Found this small icon overlapping case.

**Comment by unknowndomain** [comment]
Looking at the original PR for this there was a variant where Properties + Filters were just icons, I can't help but feel that with a tool tip these words use a lot of space up and that the purpose of this menu is quick access anyways so once you are familiar with the icons it's not an issue.

**Comment by Fenrirthviti** [comment]
Closing as all further development has moved to #2886

**Pull request closed without merging**